### ISSUE TITLE
impacts with v3.26.0

### ISSUE THREAD

**Issue by rdxmb** [projectcalico/calico#7867]
EDIT: changed description for not reproducible issue. 

## Expected Behavior

After upgrading calico from 3.25.1 to 3.26.0

* GitLab kubernetes runners continue running stable
~~* Pods behind Services with `externalTrafficPolicy: Local` continue to see the real client ip instead of showing the kubernetes node IPs.~~

## Current Behavior

- GitLab runners do not cleanup pods: https://gitlab.com/gitlab-org/gitlab-runner/-/issues/30230#note_1420409590
~~- Traefik pods running as a Deployment behind LoadBalancer-IPs (by metallb) with `externalTrafficPolicy: Local` do not see the client IP anymore~~

Downgrading to 3.25.1 again fixes both of them.

## Steps to Reproduce (for bugs)

1. Run [metallb](https://metallb.universe.tf/) for provisioning LoadBalancer IPs 
2. Create LoadBalancerIPs with `externalTrafficPolicy: Local`
3. Run pods behind that Service and check the client IPs

## Your Environment

* Calico version : 3.26.0 with `cidr: 10.244.0.0/16` , deployed by the operator
* Orchestrator version (e.g. kubernetes, mesos, rkt): kubernetes 1.24, deployed by kubeadm
* Operating System and version: Ubuntu 20.04
* metallb v0.13.10

Before calico, we had flannel running. We did not migrate with the migration project, but deleted flannel and deployed calico. So the following was our deployment process. When using the bold fields, the cluster is/was running stable.

**flannel running** -> deleting flannel -> **installing calico 3.25.1** -> deleting calico 2.35.1 -> installing calico 3.26.0 -> deleting calico 3.26.0 -> **installing calico 3.25.1**

**Comment by sridhartigera** [comment]
@rdxmb I tried using a network load balancer in aws and I can see the client IP when ```externalTrafficPolicy``` is set to ```Local```. Can you share the manifests? I can give it a shot with metalLB.

**Comment by rdxmb** [comment]
I've tried to reproduce this again in a simplified cluster to give you the manifests. Actually I do see the client IPs, so the problem seems to be another. So I will investigate this further.

Can you say something about the issue with the GitLab runners?

Kind regards!

**Comment by rdxmb** [comment]
`error killing pod: failed to "KillPodSandbox" for "12f73340-d935-465d-8b33-4fe3319e6c96" with KillPodSandboxError: "rpc error: code = Unknown desc = failed to destroy network for sandbox \"5b2d413efd6df1dfe252296175ba2b2afd784dafc423250eb3a8dcd6feca0595\": plugin type=\"calico\" failed (delete): error getting ClusterInformation: connection is unauthorized: Unauthorized"`

```
# k -n gitlab-runner-kubernetes-production get po
NAME                                             READY   STATUS        RESTARTS       AGE
gitlab-runner-864dcfb965-k6tnr                   1/1     Running       18 (28d ago)   55d
runner-ayw9xdqz-project-783-concurrent-05fzrv    0/2     Terminating   0              21m
runner-ayw9xdqz-project-783-concurrent-0wvkc7    0/2     Terminating   0              18m
runner-ayw9xdqz-project-783-concurrent-10prcmw   0/2     Terminating   0              21m
runner-ayw9xdqz-project-783-concurrent-1145m79   0/2     Terminating   0              21m
runner-ayw9xdqz-project-783-concurrent-12k9w96   0/2     Terminating   0              21m
runner-ayw9xdqz-project-783-concurrent-13rv5xv   0/2     Terminating   0              21m
runner-ayw9xdqz-project-783-concurrent-146tvj4   0/2     Terminating   0              21m
runner-ayw9xdqz-project-783-concurrent-1r45bn    0/2     Terminating   0              18m
runner-ayw9xdqz-project-783-concurrent-1tmwjp    0/2     Terminating   0              21m
runner-ayw9xdqz-project-783-concurrent-2l2pbm    0/2     Terminating   0              18m
runner-ayw9xdqz-project-783-concurrent-2ntftp    0/2     Terminating   0              21m
runner-ayw9xdqz-project-783-concurrent-3kp729    0/2     Terminating   0              17m
runner-ayw9xdqz-project-783-concurrent-3sf9pd    0/2     Terminating   0              21m
runner-ayw9xdqz-project-783-concurrent-46hw4w    0/2     Terminating   0              17m
runner-ayw9xdqz-project-783-concurrent-497sf9    0/2     Terminating   0              21m
runner-ayw9xdqz-project-783-concurrent-5qdpx8    0/2     Terminating   0              17m
runner-ayw9xdqz-project-783-concurrent-5tql8z    0/2     Terminating   0              21m
runner-ayw9xdqz-project-783-concurrent-64tw9v    0/2     Terminating   0              21m
runner-ayw9xdqz-project-783-concurrent-6dq4dz    0/2     Terminating   0              17m
runner-ayw9xdqz-project-783-concurrent-75c8k4    0/2     Terminating   0              17m
runner-ayw9xdqz-project-783-concurrent-76z4m7    0/2     Terminating   0              21m
runner-ayw9xdqz-project-783-concurrent-87hbgp    0/2     Terminating   0              21m
runner-ayw9xdqz-project-783-concurrent-8r7bcc    0/2     Terminating   0              17m
runner-ayw9xdqz-project-783-concurrent-9fxrth    0/2     Terminating   0              21m
```

**Comment by rdxmb** [comment]
client IP problem cannot be reproduced, even not in my production environment. Sorry for confusion.

**Comment by rdxmb** [comment]
@caseydavenport What do you think about the GitLab runners? Should I open a new issue for that? Kind regards.

**Comment by caseydavenport** [comment]
Sorry, misread your last comment to mean this was resolved.

**Comment by coutinhop** [comment]
@rdxmb this comment leads me to believe the root cause is a bug with the gitlab runners: https://gitlab.com/gitlab-org/gitlab-runner/-/issues/30230#note_1638579844
> Yurii Matsiuk [@ymatsiuk](https://gitlab.com/ymatsiuk) · [1 month ago](https://gitlab.com/gitlab-org/gitlab-runner/-/issues/30230#note_1638579844)
>
> I don't think it's related to signal handling. Just to clarify, the problem I'm facing is related to a typo in a config.go that ignores cleanup_grace_period_seconds if set to 0 ([toml:"cleanup_grace_period_seconds,omitzero"](https://gitlab.com/gitlab-org/gitlab-runner/-/blob/main/common/config.go?ref_type=heads#L554)), propagating GracePeriodSeconds as null to this [Delete](https://gitlab.com/gitlab-org/gitlab-runner/-/blob/main/executors/kubernetes/kubernetes.go?ref_type=heads#L1035-1040) call. Which in turn alters the logic since I want my pods to be deleted immediately (setting cleanup_grace_period_seconds=0), but instead it falls back to defaults (which presumably comes from termination grace period on the pod itself, that is set by pod_termination_grace_period_seconds=600)
>
> tl;dr: 0 should be a valid value for cleanup_grace_period_seconds according to [k8s API](https://github.com/kubernetes/apimachinery/blob/12dc3f82eb47fc972b5be0b222f4422cb1b51612/pkg/apis/meta/v1/types.go#L497-L502) and it was working in earlier versions until [this](https://gitlab.com/gitlab-org/gitlab-runner/-/merge_requests/3924/diffs#d7165d7392b1ad0c85b85647088f34bff57a219a_474_473)

Could you please verify that the pods clean up after the default 600 second grace period? (Also, it's not quite clear to me why downgrading the calico version impacts that, unless the gitlab runners' version depends on calico version, and the one corresponding to calico v3.25.1 does not contain this bug). Please do correct me if I'm wrong, though.

**Comment by coutinhop** [comment]
@rdxmb I'll close the issue, feel free to reach out if you're still seeing this problem (and please provide more info if so).
